# Loading Data into Pandas

In [15]:
import pandas as pd

#reading csv
df = pd.read_csv('pokemon_data.csv')
print(df.head())

#reading excel
#df_xlsx = pd.read_excel('pokemon_data.xlsx')
#print(df_xlsx.head())

#reading txt
#df_txt = pd.read_csv('pokemon_data.txt', delimiter = '\t')
#print(df_txt.head())

   #                   Name Type 1  Type 2  HP  Attack  Defense  Sp. Atk  \
0  1              Bulbasaur  Grass  Poison  45      49       49       65   
1  2                Ivysaur  Grass  Poison  60      62       63       80   
2  3               Venusaur  Grass  Poison  80      82       83      100   
3  3  VenusaurMega Venusaur  Grass  Poison  80     100      123      122   
4  4             Charmander   Fire     NaN  39      52       43       60   

   Sp. Def  Speed  Generation  Legendary  
0       65     45           1      False  
1       80     60           1      False  
2      100     80           1      False  
3      120     80           1      False  
4       50     65           1      False  


# Reading Data in Pandas

In [20]:
#Read Headers
#print(df.columns)
#or 
df.columns

Index(['#', 'Name', 'Type 1', 'Type 2', 'HP', 'Attack', 'Defense', 'Sp. Atk',
       'Sp. Def', 'Speed', 'Generation', 'Legendary'],
      dtype='object')

In [24]:
#Read each Column
print(df[['Name', 'Type 1', 'HP']][0:10])

                        Name Type 1  HP
0                  Bulbasaur  Grass  45
1                    Ivysaur  Grass  60
2                   Venusaur  Grass  80
3      VenusaurMega Venusaur  Grass  80
4                 Charmander   Fire  39
5                 Charmeleon   Fire  58
6                  Charizard   Fire  78
7  CharizardMega Charizard X   Fire  78
8  CharizardMega Charizard Y   Fire  78
9                   Squirtle  Water  44


In [27]:
#Read each Row
print(df.iloc[0:4])

   #                   Name Type 1  Type 2  HP  Attack  Defense  Sp. Atk  \
0  1              Bulbasaur  Grass  Poison  45      49       49       65   
1  2                Ivysaur  Grass  Poison  60      62       63       80   
2  3               Venusaur  Grass  Poison  80      82       83      100   
3  3  VenusaurMega Venusaur  Grass  Poison  80     100      123      122   

   Sp. Def  Speed  Generation  Legendary  
0       65     45           1      False  
1       80     60           1      False  
2      100     80           1      False  
3      120     80           1      False  


In [37]:
#For getting a list of a column
for index, row in df.iterrows():
    print(index, row['Name'])

0 Bulbasaur
1 Ivysaur
2 Venusaur
3 VenusaurMega Venusaur
4 Charmander
5 Charmeleon
6 Charizard
7 CharizardMega Charizard X
8 CharizardMega Charizard Y
9 Squirtle
10 Wartortle
11 Blastoise
12 BlastoiseMega Blastoise
13 Caterpie
14 Metapod
15 Butterfree
16 Weedle
17 Kakuna
18 Beedrill
19 BeedrillMega Beedrill
20 Pidgey
21 Pidgeotto
22 Pidgeot
23 PidgeotMega Pidgeot
24 Rattata
25 Raticate
26 Spearow
27 Fearow
28 Ekans
29 Arbok
30 Pikachu
31 Raichu
32 Sandshrew
33 Sandslash
34 Nidoran (Female)
35 Nidorina
36 Nidoqueen
37 Nidoran (Male)
38 Nidorino
39 Nidoking
40 Clefairy
41 Clefable
42 Vulpix
43 Ninetales
44 Jigglypuff
45 Wigglytuff
46 Zubat
47 Golbat
48 Oddish
49 Gloom
50 Vileplume
51 Paras
52 Parasect
53 Venonat
54 Venomoth
55 Diglett
56 Dugtrio
57 Meowth
58 Persian
59 Psyduck
60 Golduck
61 Mankey
62 Primeape
63 Growlithe
64 Arcanine
65 Poliwag
66 Poliwhirl
67 Poliwrath
68 Abra
69 Kadabra
70 Alakazam
71 AlakazamMega Alakazam
72 Machop
73 Machoke
74 Machamp
75 Bellsprout
76 Weepinbell
77 

746 MeowsticMale
747 MeowsticFemale
748 Honedge
749 Doublade
750 AegislashBlade Forme
751 AegislashShield Forme
752 Spritzee
753 Aromatisse
754 Swirlix
755 Slurpuff
756 Inkay
757 Malamar
758 Binacle
759 Barbaracle
760 Skrelp
761 Dragalge
762 Clauncher
763 Clawitzer
764 Helioptile
765 Heliolisk
766 Tyrunt
767 Tyrantrum
768 Amaura
769 Aurorus
770 Sylveon
771 Hawlucha
772 Dedenne
773 Carbink
774 Goomy
775 Sliggoo
776 Goodra
777 Klefki
778 Phantump
779 Trevenant
780 PumpkabooAverage Size
781 PumpkabooSmall Size
782 PumpkabooLarge Size
783 PumpkabooSuper Size
784 GourgeistAverage Size
785 GourgeistSmall Size
786 GourgeistLarge Size
787 GourgeistSuper Size
788 Bergmite
789 Avalugg
790 Noibat
791 Noivern
792 Xerneas
793 Yveltal
794 Zygarde50% Forme
795 Diancie
796 DiancieMega Diancie
797 HoopaHoopa Confined
798 HoopaHoopa Unbound
799 Volcanion


In [29]:
#Read a specific location (R, C)
print(df.iloc[2, 1])

Venusaur


In [36]:
df.loc[df['Type 1'] == 'Fire'][0:10]

#                       Name Type 1  Type 2  HP  Attack  Defense  \
4    4                 Charmander   Fire     NaN  39      52       43   
5    5                 Charmeleon   Fire     NaN  58      64       58   
6    6                  Charizard   Fire  Flying  78      84       78   
7    6  CharizardMega Charizard X   Fire  Dragon  78     130      111   
8    6  CharizardMega Charizard Y   Fire  Flying  78     104       78   
42  37                     Vulpix   Fire     NaN  38      41       40   
43  38                  Ninetales   Fire     NaN  73      76       75   
63  58                  Growlithe   Fire     NaN  55      70       45   
64  59                   Arcanine   Fire     NaN  90     110       80   
83  77                     Ponyta   Fire     NaN  50      85       55   

    Sp. Atk  Sp. Def  Speed  Generation  Legendary  
4        60       50     65           1      False  
5        80       65     80           1      False  
6       109       85    100           1      False  
7       130       85    100           1      False  
8       159      115    100           1      False  
42       50       65     65           1      False  
43       81      100    100           1      False  
63       70       50     60           1      False  
64      100       80     95           1      False  
83       65       65     90           1      False

In [35]:
df.loc[df['Type 1'] == 'Grass'][0:10]

#                   Name Type 1  Type 2  HP  Attack  Defense  Sp. Atk  \
0    1              Bulbasaur  Grass  Poison  45      49       49       65   
1    2                Ivysaur  Grass  Poison  60      62       63       80   
2    3               Venusaur  Grass  Poison  80      82       83      100   
3    3  VenusaurMega Venusaur  Grass  Poison  80     100      123      122   
48  43                 Oddish  Grass  Poison  45      50       55       75   
49  44                  Gloom  Grass  Poison  60      65       70       85   
50  45              Vileplume  Grass  Poison  75      80       85      110   
75  69             Bellsprout  Grass  Poison  50      75       35       70   
76  70             Weepinbell  Grass  Poison  65      90       50       85   
77  71             Victreebel  Grass  Poison  80     105       65      100   

    Sp. Def  Speed  Generation  Legendary  
0        65     45           1      False  
1        80     60           1      False  
2       100     80           1      False  
3       120     80           1      False  
48       65     30           1      False  
49       75     40           1      False  
50       90     50           1      False  
75       30     40           1      False  
76       45     55           1      False  
77       70     70           1      False

# Sorting/Describing Data

In [38]:
df.describe()

#          HP      Attack     Defense     Sp. Atk     Sp. Def  \
count  800.000000  800.000000  800.000000  800.000000  800.000000  800.000000   
mean   362.813750   69.258750   79.001250   73.842500   72.820000   71.902500   
std    208.343798   25.534669   32.457366   31.183501   32.722294   27.828916   
min      1.000000    1.000000    5.000000    5.000000   10.000000   20.000000   
25%    184.750000   50.000000   55.000000   50.000000   49.750000   50.000000   
50%    364.500000   65.000000   75.000000   70.000000   65.000000   70.000000   
75%    539.250000   80.000000  100.000000   90.000000   95.000000   90.000000   
max    721.000000  255.000000  190.000000  230.000000  194.000000  230.000000   

            Speed  Generation  
count  800.000000   800.00000  
mean    68.277500     3.32375  
std     29.060474     1.66129  
min      5.000000     1.00000  
25%     45.000000     2.00000  
50%     65.000000     3.00000  
75%     90.000000     5.00000  
max    180.000000     6.00000

In [40]:
df.sort_values('Name')

#                     Name   Type 1  Type 2   HP  Attack  Defense  \
510  460                Abomasnow    Grass     Ice   90      92       75   
511  460  AbomasnowMega Abomasnow    Grass     Ice   90     132      105   
68    63                     Abra  Psychic     NaN   25      20       15   
392  359                    Absol     Dark     NaN   65     130       60   
393  359          AbsolMega Absol     Dark     NaN   65     150       60   
..   ...                      ...      ...     ...  ...     ...      ...   
632  571                  Zoroark     Dark     NaN   60     105       60   
631  570                    Zorua     Dark     NaN   40      65       40   
46    41                    Zubat   Poison  Flying   40      45       35   
695  634                 Zweilous     Dark  Dragon   72      85       70   
794  718         Zygarde50% Forme   Dragon  Ground  108     100      121   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
510       92       85     60           4      False  
511      132      105     30           4      False  
68       105       55     90           1      False  
392       75       60     75           3      False  
393      115       60    115           3      False  
..       ...      ...    ...         ...        ...  
632      120       60    105           5      False  
631       80       40     65           5      False  
46        30       40     55           1      False  
695       65       70     58           5      False  
794       81       95     95           6       True  

[800 rows x 12 columns]

In [41]:
df.sort_values('Name', ascending = False)

#                     Name   Type 1  Type 2   HP  Attack  Defense  \
794  718         Zygarde50% Forme   Dragon  Ground  108     100      121   
695  634                 Zweilous     Dark  Dragon   72      85       70   
46    41                    Zubat   Poison  Flying   40      45       35   
631  570                    Zorua     Dark     NaN   40      65       40   
632  571                  Zoroark     Dark     NaN   60     105       60   
..   ...                      ...      ...     ...  ...     ...      ...   
393  359          AbsolMega Absol     Dark     NaN   65     150       60   
392  359                    Absol     Dark     NaN   65     130       60   
68    63                     Abra  Psychic     NaN   25      20       15   
511  460  AbomasnowMega Abomasnow    Grass     Ice   90     132      105   
510  460                Abomasnow    Grass     Ice   90      92       75   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
794       81       95     95           6       True  
695       65       70     58           5      False  
46        30       40     55           1      False  
631       80       40     65           5      False  
632      120       60    105           5      False  
..       ...      ...    ...         ...        ...  
393      115       60    115           3      False  
392       75       60     75           3      False  
68       105       55     90           1      False  
511      132      105     30           4      False  
510       92       85     60           4      False  

[800 rows x 12 columns]

In [44]:
df.sort_values(['Type 1', 'HP'], ascending = [1, 0])

#                     Name Type 1    Type 2  HP  Attack  Defense  \
139  129                 Magikarp  Water       NaN  20      10       55   
381  349                   Feebas  Water       NaN  20      15       20   
97    90                 Shellder  Water       NaN  30      65      100   
106   98                   Krabby  Water       NaN  30     105       90   
125  116                   Horsea  Water       NaN  30      40       70   
..   ...                      ...    ...       ...  ..     ...      ...   
232  214  HeracrossMega Heracross    Bug  Fighting  80     185      115   
678  617                 Accelgor    Bug       NaN  80      70       40   
734  666                 Vivillon    Bug    Flying  80      52       50   
698  637                Volcarona    Bug      Fire  85      60       65   
520  469                  Yanmega    Bug    Flying  86      76       86   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
139       15       20     80           1      False  
381       10       55     80           3      False  
97        45       25     40           1      False  
106       25       25     50           1      False  
125       70       25     60           1      False  
..       ...      ...    ...         ...        ...  
232       40      105     75           2      False  
678      100       60    145           5      False  
734       90       50     89           6      False  
698      135      105    100           5      False  
520      116       56     95           4      False  

[800 rows x 12 columns]

- **1 = True, 0 = False**

# Making changes to the data

### Adding Columns to create a new Column

- **Method 1**

In [71]:
df['Total'] = df['HP'] + df['Attack'] + df['Defense'] + df['Sp. Atk'] + df['Sp. Def'] + df['Speed']

df.head()

#                   Name Type 1  Type 2  Generation  Legendary  HP  Attack  \
0  1              Bulbasaur  Grass  Poison           1      False  45      49   
1  2                Ivysaur  Grass  Poison           1      False  60      62   
2  3               Venusaur  Grass  Poison           1      False  80      82   
3  3  VenusaurMega Venusaur  Grass  Poison           1      False  80     100   
4  4             Charmander   Fire     NaN           1      False  39      52   

   Defense  Sp. Atk  Sp. Def  Speed  Total  
0       49       65       65     45    318  
1       63       80       80     60    405  
2       83      100      100     80    525  
3      123      122      120     80    625  
4       43       60       50     65    309

- **Method 2**

In [70]:
df['Total'] = df.iloc[:, 4:12].sum(axis=1)

df.head(4)

#                   Name Type 1  Type 2  Generation  Legendary  HP  Attack  \
0  1              Bulbasaur  Grass  Poison           1      False  45      49   
1  2                Ivysaur  Grass  Poison           1      False  60      62   
2  3               Venusaur  Grass  Poison           1      False  80      82   
3  3  VenusaurMega Venusaur  Grass  Poison           1      False  80     100   

   Defense  Sp. Atk  Sp. Def  Speed  Total  
0       49       65       65     45  319.0  
1       63       80       80     60  406.0  
2       83      100      100     80  526.0  
3      123      122      120     80  626.0

****Sum is 319 instead of 318 because the rows are rearranged therefore adding Legendary column too*

- **: means all rows are included**

- **axis = 1 means horizontal summing (for vertical summing axis = 0)**

### Rearranging Columns

In [57]:
cols = list(df.columns)
df = df[cols[0:4] + [cols[-1]]+cols[4:12]]

df.head()

#                   Name Type 1  Type 2  Generation  Legendary  Total  HP  \
0  1              Bulbasaur  Grass  Poison           1      False    318  45   
1  2                Ivysaur  Grass  Poison           1      False    405  60   
2  3               Venusaur  Grass  Poison           1      False    525  80   
3  3  VenusaurMega Venusaur  Grass  Poison           1      False    625  80   
4  4             Charmander   Fire     NaN           1      False    309  39   

   Attack  Defense  Sp. Atk  Sp. Def  Speed  
0      49       49       65       65     45  
1      62       63       80       80     60  
2      82       83      100      100     80  
3     100      123      122      120     80  
4      52       43       60       50     65

### Deleting Columns

In [65]:
df = df.drop(columns=['Total'])
             
df.head()

#                   Name Type 1  Type 2  Generation  Legendary  HP  Attack  \
0  1              Bulbasaur  Grass  Poison           1      False  45      49   
1  2                Ivysaur  Grass  Poison           1      False  60      62   
2  3               Venusaur  Grass  Poison           1      False  80      82   
3  3  VenusaurMega Venusaur  Grass  Poison           1      False  80     100   
4  4             Charmander   Fire     NaN           1      False  39      52   

   Defense  Sp. Atk  Sp. Def  Speed  
0       49       65       65     45  
1       63       80       80     60  
2       83      100      100     80  
3      123      122      120     80  
4       43       60       50     65

# Saving our Data (Exporting into Desired Format)

In [73]:
Saving as CSV
df.to_csv('modified_pokemon.csv', index = False)

#Saving as Excel
#df.to_excel('modified_pokemon.csv', index = False)

#Saving as TXT
#df.to_csv('modified_pokemon.csv', index = False, sep = '\t')

# Filtering Data

In [83]:
df.loc[(df['Type 1'] == 'Water') & (df['Type 2'] == 'Ice')]

#      Name Type 1 Type 2  Generation  Legendary   HP  Attack  Defense  \
94    87   Dewgong  Water    Ice           1      False   90      70       80   
98    91  Cloyster  Water    Ice           1      False   50      95      180   
142  131    Lapras  Water    Ice           1      False  130      85       80   

     Sp. Atk  Sp. Def  Speed  Total  
94        70       95     70    475  
98        85       45     70    525  
142       85       95     60    535

In [84]:
df.loc[(df['Type 1'] == 'Water') | (df['Type 2'] == 'Ice')]

#                     Name Type 1 Type 2  Generation  Legendary   HP  \
9      7                 Squirtle  Water    NaN           1      False   44   
10     8                Wartortle  Water    NaN           1      False   59   
11     9                Blastoise  Water    NaN           1      False   79   
12     9  BlastoiseMega Blastoise  Water    NaN           1      False   79   
59    54                  Psyduck  Water    NaN           1      False   50   
..   ...                      ...    ...    ...         ...        ...  ...   
726  658                 Greninja  Water   Dark           6      False   72   
762  692                Clauncher  Water    NaN           6      False   50   
763  693                Clawitzer  Water    NaN           6      False   71   
768  698                   Amaura   Rock    Ice           6      False   77   
769  699                  Aurorus   Rock    Ice           6      False  123   

     Attack  Defense  Sp. Atk  Sp. Def  Speed  Total  
9        48       65       50       64     43    314  
10       63       80       65       80     58    405  
11       83      100       85      105     78    530  
12      103      120      135      115     78    630  
59       52       48       65       50     55    320  
..      ...      ...      ...      ...    ...    ...  
726      95       67      103       71    122    530  
762      53       62       58       63     44    330  
763      73       88      120       89     59    500  
768      59       50       67       63     46    362  
769      77       72       99       92     58    521  

[123 rows x 13 columns]

In [114]:
new_df = df.loc[(df['Type 1'] == 'Ice') | (df['Speed'] > 50) & (df['Type 2'] == 'NaN' )]

new_df = new_df.reset_index(drop = True)

print(new_df)

      #               Name Type 1   Type 2  Generation  Legendary   HP  \
0   124               Jynx    Ice  Psychic           1      False   65   
1   144           Articuno    Ice   Flying           1       True   90   
2   220             Swinub    Ice   Ground           2      False   50   
3   221          Piloswine    Ice   Ground           2      False  100   
4   225           Delibird    Ice   Flying           2      False   45   
5   238           Smoochum    Ice  Psychic           2      False   45   
6   361            Snorunt    Ice      NaN           3      False   50   
7   362             Glalie    Ice      NaN           3      False   80   
8   362  GlalieMega Glalie    Ice      NaN           3      False   80   
9   363             Spheal    Ice    Water           3      False   70   
10  364             Sealeo    Ice    Water           3      False   90   
11  365            Walrein    Ice    Water           3      False  110   
12  378             Regice    Ice     

- **Reset_index = starts a new index from 1**

- **drop = removes the old index from the data frame**

In [117]:
df.loc[~df['Name'].str.contains('Mega')]

#                 Name   Type 1  Type 2  Generation  Legendary   HP  \
0      1            Bulbasaur    Grass  Poison           1      False   45   
1      2              Ivysaur    Grass  Poison           1      False   60   
2      3             Venusaur    Grass  Poison           1      False   80   
4      4           Charmander     Fire     NaN           1      False   39   
5      5           Charmeleon     Fire     NaN           1      False   58   
..   ...                  ...      ...     ...         ...        ...  ...   
794  718     Zygarde50% Forme   Dragon  Ground           6       True  108   
795  719              Diancie     Rock   Fairy           6       True   50   
797  720  HoopaHoopa Confined  Psychic   Ghost           6       True   80   
798  720   HoopaHoopa Unbound  Psychic    Dark           6       True   80   
799  721            Volcanion     Fire   Water           6       True   80   

     Attack  Defense  Sp. Atk  Sp. Def  Speed  Total  
0        49       49       65       65     45    318  
1        62       63       80       80     60    405  
2        82       83      100      100     80    525  
4        52       43       60       50     65    309  
5        64       58       80       65     80    405  
..      ...      ...      ...      ...    ...    ...  
794     100      121       81       95     95    600  
795     100      150      100      150     50    600  
797     110       60      150      130     70    600  
798     160       60      170      130     80    680  
799     110      120      130       90     70    600  

[751 rows x 13 columns]

- **~ means removal of the filtered data after it's usage**

# Regex Filtering

In [121]:
import re

df.loc[df['Type 1'].str.contains('Fire|Grass', regex = True)]

#                   Name Type 1  Type 2  Generation  Legendary   HP  \
0      1              Bulbasaur  Grass  Poison           1      False   45   
1      2                Ivysaur  Grass  Poison           1      False   60   
2      3               Venusaur  Grass  Poison           1      False   80   
3      3  VenusaurMega Venusaur  Grass  Poison           1      False   80   
4      4             Charmander   Fire     NaN           1      False   39   
..   ...                    ...    ...     ...         ...        ...  ...   
735  667                 Litleo   Fire  Normal           6      False   62   
736  668                 Pyroar   Fire  Normal           6      False   86   
740  672                 Skiddo  Grass     NaN           6      False   66   
741  673                 Gogoat  Grass     NaN           6      False  123   
799  721              Volcanion   Fire   Water           6       True   80   

     Attack  Defense  Sp. Atk  Sp. Def  Speed  Total  
0        49       49       65       65     45    318  
1        62       63       80       80     60    405  
2        82       83      100      100     80    525  
3       100      123      122      120     80    625  
4        52       43       60       50     65    309  
..      ...      ...      ...      ...    ...    ...  
735      50       58       73       54     72    369  
736      68       72      109       66    106    507  
740      65       48       62       57     52    350  
741     100       62       97       81     68    531  
799     110      120      130       90     70    600  

[122 rows x 13 columns]

- **regex means that regular expressions are used in the script**

- **'Fire|Grass' is very Case-sensetive**

In [122]:
df.loc[df['Type 1'].str.contains('fire|grass', flags = re.I, regex = True)]

#                   Name Type 1  Type 2  Generation  Legendary   HP  \
0      1              Bulbasaur  Grass  Poison           1      False   45   
1      2                Ivysaur  Grass  Poison           1      False   60   
2      3               Venusaur  Grass  Poison           1      False   80   
3      3  VenusaurMega Venusaur  Grass  Poison           1      False   80   
4      4             Charmander   Fire     NaN           1      False   39   
..   ...                    ...    ...     ...         ...        ...  ...   
735  667                 Litleo   Fire  Normal           6      False   62   
736  668                 Pyroar   Fire  Normal           6      False   86   
740  672                 Skiddo  Grass     NaN           6      False   66   
741  673                 Gogoat  Grass     NaN           6      False  123   
799  721              Volcanion   Fire   Water           6       True   80   

     Attack  Defense  Sp. Atk  Sp. Def  Speed  Total  
0        49       49       65       65     45    318  
1        62       63       80       80     60    405  
2        82       83      100      100     80    525  
3       100      123      122      120     80    625  
4        52       43       60       50     65    309  
..      ...      ...      ...      ...    ...    ...  
735      50       58       73       54     72    369  
736      68       72      109       66    106    507  
740      65       48       62       57     52    350  
741     100       62       97       81     68    531  
799     110      120      130       90     70    600  

[122 rows x 13 columns]

- **'flags = re.I' ignores the Case of the Data**

In [125]:
df.loc[df['Name'].str.contains('^pi[a-z]*', flags = re.I, regex = True)]

#                 Name    Type 1    Type 2  Generation  Legendary   HP  \
20    16               Pidgey    Normal    Flying           1      False   40   
21    17            Pidgeotto    Normal    Flying           1      False   63   
22    18              Pidgeot    Normal    Flying           1      False   83   
23    18  PidgeotMega Pidgeot    Normal    Flying           1      False   83   
30    25              Pikachu  Electric       NaN           1      False   35   
136  127               Pinsir       Bug       NaN           1      False   65   
137  127    PinsirMega Pinsir       Bug    Flying           1      False   65   
186  172                Pichu  Electric       NaN           2      False   20   
219  204               Pineco       Bug       NaN           2      False   50   
239  221            Piloswine       Ice    Ground           2      False  100   
438  393               Piplup     Water       NaN           4      False   53   
558  499              Pignite      Fire  Fighting           5      False   90   
578  519               Pidove    Normal    Flying           5      False   50   

     Attack  Defense  Sp. Atk  Sp. Def  Speed  Total  
20       45       40       35       35     56    251  
21       60       55       50       50     71    349  
22       80       75       70       70    101    479  
23       80       80      135       80    121    579  
30       55       40       50       50     90    320  
136     125      100       55       70     85    500  
137     155      120       65       90    105    600  
186      40       15       35       35     60    205  
219      65       90       35       35     15    290  
239     100       80       60       60     50    450  
438      51       53       61       56     40    314  
558      93       55       70       55     55    418  
578      55       50       36       30     43    264

# Conditional Changes

In [132]:
df.loc[df['Type 1'] == 'Fire', 'Type 1'] = 'Flamer'

df

#                   Name   Type 1  Type 2  Generation  Legendary  HP  \
0      1              Bulbasaur    Grass  Poison           1      False  45   
1      2                Ivysaur    Grass  Poison           1      False  60   
2      3               Venusaur    Grass  Poison           1      False  80   
3      3  VenusaurMega Venusaur    Grass  Poison           1      False  80   
4      4             Charmander   Flamer     NaN           1      False  39   
..   ...                    ...      ...     ...         ...        ...  ..   
795  719                Diancie     Rock   Fairy           6       True  50   
796  719    DiancieMega Diancie     Rock   Fairy           6       True  50   
797  720    HoopaHoopa Confined  Psychic   Ghost           6       True  80   
798  720     HoopaHoopa Unbound  Psychic    Dark           6       True  80   
799  721              Volcanion   Flamer   Water           6       True  80   

     Attack  Defense  Sp. Atk  Sp. Def  Speed  Total  
0        49       49       65       65     45    318  
1        62       63       80       80     60    405  
2        82       83      100      100     80    525  
3       100      123      122      120     80    625  
4        52       43       60       50     65    309  
..      ...      ...      ...      ...    ...    ...  
795     100      150      100      150     50    600  
796     160      110      160      110    110    700  
797     110       60      150      130     70    600  
798     160       60      170      130     80    680  
799     110      120      130       90     70    600  

[800 rows x 13 columns]

- **This script changes Fire into Flamer and change it back to Fire, we write the following code:**

In [133]:
df.loc[df['Type 1'] == 'Flamer', 'Type 1'] = 'Fire'

df

#                   Name   Type 1  Type 2  Generation  Legendary  HP  \
0      1              Bulbasaur    Grass  Poison           1      False  45   
1      2                Ivysaur    Grass  Poison           1      False  60   
2      3               Venusaur    Grass  Poison           1      False  80   
3      3  VenusaurMega Venusaur    Grass  Poison           1      False  80   
4      4             Charmander     Fire     NaN           1      False  39   
..   ...                    ...      ...     ...         ...        ...  ..   
795  719                Diancie     Rock   Fairy           6       True  50   
796  719    DiancieMega Diancie     Rock   Fairy           6       True  50   
797  720    HoopaHoopa Confined  Psychic   Ghost           6       True  80   
798  720     HoopaHoopa Unbound  Psychic    Dark           6       True  80   
799  721              Volcanion     Fire   Water           6       True  80   

     Attack  Defense  Sp. Atk  Sp. Def  Speed  Total  
0        49       49       65       65     45    318  
1        62       63       80       80     60    405  
2        82       83      100      100     80    525  
3       100      123      122      120     80    625  
4        52       43       60       50     65    309  
..      ...      ...      ...      ...    ...    ...  
795     100      150      100      150     50    600  
796     160      110      160      110    110    700  
797     110       60      150      130     70    600  
798     160       60      170      130     80    680  
799     110      120      130       90     70    600  

[800 rows x 13 columns]

In [140]:
df.loc[df['Total'] > 500, ['Generation', 'Legendary']] = ['Good', 'Great']

df

#                   Name   Type 1  Type 2 Generation Legendary  HP  \
0      1              Bulbasaur    Grass  Poison          1     False  45   
1      2                Ivysaur    Grass  Poison          1     False  60   
2      3               Venusaur    Grass  Poison       Good     Great  80   
3      3  VenusaurMega Venusaur    Grass  Poison       Good     Great  80   
4      4             Charmander     Fire     NaN          1     False  39   
..   ...                    ...      ...     ...        ...       ...  ..   
795  719                Diancie     Rock   Fairy       Good     Great  50   
796  719    DiancieMega Diancie     Rock   Fairy       Good     Great  50   
797  720    HoopaHoopa Confined  Psychic   Ghost       Good     Great  80   
798  720     HoopaHoopa Unbound  Psychic    Dark       Good     Great  80   
799  721              Volcanion     Fire   Water       Good     Great  80   

     Attack  Defense  Sp. Atk  Sp. Def  Speed  Total  
0        49       49       65       65     45    318  
1        62       63       80       80     60    405  
2        82       83      100      100     80    525  
3       100      123      122      120     80    625  
4        52       43       60       50     65    309  
..      ...      ...      ...      ...    ...    ...  
795     100      150      100      150     50    600  
796     160      110      160      110    110    700  
797     110       60      150      130     70    600  
798     160       60      170      130     80    680  
799     110      120      130       90     70    600  

[800 rows x 13 columns]

# Aggregated Statistics (Group By)

In [146]:
df.groupby(['Type 1']).mean().sort_values('HP', ascending = False)

#  Generation  Legendary         HP      Attack  \
Type 1                                                               
Dragon    474.375000    3.875000   0.375000  83.312500  112.125000   
Normal    319.173469    3.051020   0.020408  77.275510   73.469388   
Fairy     449.529412    4.117647   0.058824  74.117647   61.529412   
Ground    356.281250    3.156250   0.125000  73.781250   95.750000   
Water     303.089286    2.857143   0.035714  72.062500   74.151786   
Ice       423.541667    3.541667   0.083333  72.000000   72.750000   
Flying    677.750000    5.500000   0.500000  70.750000   78.750000   
Psychic   380.807018    3.385965   0.245614  70.631579   71.456140   
Fire      327.403846    3.211538   0.096154  69.903846   84.769231   
Fighting  363.851852    3.370370   0.000000  69.851852   96.777778   
Grass     344.871429    3.357143   0.042857  67.271429   73.214286   
Poison    251.785714    2.535714   0.000000  67.250000   74.678571   
Dark      461.354839    4.032258   0.064516  66.806452   88.387097   
Rock      392.727273    3.454545   0.090909  65.363636   92.863636   
Steel     442.851852    3.851852   0.148148  65.222222   92.703704   
Ghost     486.500000    4.187500   0.062500  64.437500   73.781250   
Electric  363.500000    3.272727   0.090909  59.795455   69.090909   
Bug       334.492754    3.217391   0.000000  56.884058   70.971014   

             Defense    Sp. Atk    Sp. Def       Speed       Total  
Type 1                                                              
Dragon     86.375000  96.843750  88.843750   83.031250  550.531250  
Normal     59.846939  55.816327  63.724490   71.551020  401.683673  
Fairy      65.705882  78.529412  84.705882   48.588235  413.176471  
Ground     84.843750  56.468750  62.750000   63.906250  437.500000  
Water      72.946429  74.812500  70.517857   65.964286  430.455357  
Ice        71.416667  77.541667  76.291667   63.458333  433.458333  
Flying     66.250000  94.250000  72.500000  102.500000  485.000000  
Psychic    67.684211  98.403509  86.280702   81.491228  475.947368  
Fire       67.769231  88.980769  72.211538   74.442308  458.076923  
Fighting   65.925926  53.111111  64.703704   66.074074  416.444444  
Grass      70.800000  77.500000  70.428571   61.928571  421.142857  
Poison     68.821429  60.428571  64.392857   63.571429  399.142857  
Dark       70.225806  74.645161  69.516129   76.161290  445.741935  
Rock      100.795455  63.340909  75.477273   55.909091  453.750000  
Steel     126.370370  67.518519  80.629630   55.259259  487.703704  
Ghost      81.187500  79.343750  76.468750   64.343750  439.562500  
Electric   66.295455  90.022727  73.704545   84.500000  443.409091  
Bug        70.724638  53.869565  64.797101   61.681159  378.927536

In [148]:
df.groupby(['Type 1']).mean().sort_values('Defense', ascending = False)

#  Generation  Legendary         HP      Attack  \
Type 1                                                               
Steel     442.851852    3.851852   0.148148  65.222222   92.703704   
Rock      392.727273    3.454545   0.090909  65.363636   92.863636   
Dragon    474.375000    3.875000   0.375000  83.312500  112.125000   
Ground    356.281250    3.156250   0.125000  73.781250   95.750000   
Ghost     486.500000    4.187500   0.062500  64.437500   73.781250   
Water     303.089286    2.857143   0.035714  72.062500   74.151786   
Ice       423.541667    3.541667   0.083333  72.000000   72.750000   
Grass     344.871429    3.357143   0.042857  67.271429   73.214286   
Bug       334.492754    3.217391   0.000000  56.884058   70.971014   
Dark      461.354839    4.032258   0.064516  66.806452   88.387097   
Poison    251.785714    2.535714   0.000000  67.250000   74.678571   
Fire      327.403846    3.211538   0.096154  69.903846   84.769231   
Psychic   380.807018    3.385965   0.245614  70.631579   71.456140   
Electric  363.500000    3.272727   0.090909  59.795455   69.090909   
Flying    677.750000    5.500000   0.500000  70.750000   78.750000   
Fighting  363.851852    3.370370   0.000000  69.851852   96.777778   
Fairy     449.529412    4.117647   0.058824  74.117647   61.529412   
Normal    319.173469    3.051020   0.020408  77.275510   73.469388   

             Defense    Sp. Atk    Sp. Def       Speed       Total  
Type 1                                                              
Steel     126.370370  67.518519  80.629630   55.259259  487.703704  
Rock      100.795455  63.340909  75.477273   55.909091  453.750000  
Dragon     86.375000  96.843750  88.843750   83.031250  550.531250  
Ground     84.843750  56.468750  62.750000   63.906250  437.500000  
Ghost      81.187500  79.343750  76.468750   64.343750  439.562500  
Water      72.946429  74.812500  70.517857   65.964286  430.455357  
Ice        71.416667  77.541667  76.291667   63.458333  433.458333  
Grass      70.800000  77.500000  70.428571   61.928571  421.142857  
Bug        70.724638  53.869565  64.797101   61.681159  378.927536  
Dark       70.225806  74.645161  69.516129   76.161290  445.741935  
Poison     68.821429  60.428571  64.392857   63.571429  399.142857  
Fire       67.769231  88.980769  72.211538   74.442308  458.076923  
Psychic    67.684211  98.403509  86.280702   81.491228  475.947368  
Electric   66.295455  90.022727  73.704545   84.500000  443.409091  
Flying     66.250000  94.250000  72.500000  102.500000  485.000000  
Fighting   65.925926  53.111111  64.703704   66.074074  416.444444  
Fairy      65.705882  78.529412  84.705882   48.588235  413.176471  
Normal     59.846939  55.816327  63.724490   71.551020  401.683673

In [147]:
df.groupby(['Type 1']).mean().sort_values('Attack', ascending = False)

#  Generation  Legendary         HP      Attack  \
Type 1                                                               
Dragon    474.375000    3.875000   0.375000  83.312500  112.125000   
Fighting  363.851852    3.370370   0.000000  69.851852   96.777778   
Ground    356.281250    3.156250   0.125000  73.781250   95.750000   
Rock      392.727273    3.454545   0.090909  65.363636   92.863636   
Steel     442.851852    3.851852   0.148148  65.222222   92.703704   
Dark      461.354839    4.032258   0.064516  66.806452   88.387097   
Fire      327.403846    3.211538   0.096154  69.903846   84.769231   
Flying    677.750000    5.500000   0.500000  70.750000   78.750000   
Poison    251.785714    2.535714   0.000000  67.250000   74.678571   
Water     303.089286    2.857143   0.035714  72.062500   74.151786   
Ghost     486.500000    4.187500   0.062500  64.437500   73.781250   
Normal    319.173469    3.051020   0.020408  77.275510   73.469388   
Grass     344.871429    3.357143   0.042857  67.271429   73.214286   
Ice       423.541667    3.541667   0.083333  72.000000   72.750000   
Psychic   380.807018    3.385965   0.245614  70.631579   71.456140   
Bug       334.492754    3.217391   0.000000  56.884058   70.971014   
Electric  363.500000    3.272727   0.090909  59.795455   69.090909   
Fairy     449.529412    4.117647   0.058824  74.117647   61.529412   

             Defense    Sp. Atk    Sp. Def       Speed       Total  
Type 1                                                              
Dragon     86.375000  96.843750  88.843750   83.031250  550.531250  
Fighting   65.925926  53.111111  64.703704   66.074074  416.444444  
Ground     84.843750  56.468750  62.750000   63.906250  437.500000  
Rock      100.795455  63.340909  75.477273   55.909091  453.750000  
Steel     126.370370  67.518519  80.629630   55.259259  487.703704  
Dark       70.225806  74.645161  69.516129   76.161290  445.741935  
Fire       67.769231  88.980769  72.211538   74.442308  458.076923  
Flying     66.250000  94.250000  72.500000  102.500000  485.000000  
Poison     68.821429  60.428571  64.392857   63.571429  399.142857  
Water      72.946429  74.812500  70.517857   65.964286  430.455357  
Ghost      81.187500  79.343750  76.468750   64.343750  439.562500  
Normal     59.846939  55.816327  63.724490   71.551020  401.683673  
Grass      70.800000  77.500000  70.428571   61.928571  421.142857  
Ice        71.416667  77.541667  76.291667   63.458333  433.458333  
Psychic    67.684211  98.403509  86.280702   81.491228  475.947368  
Bug        70.724638  53.869565  64.797101   61.681159  378.927536  
Electric   66.295455  90.022727  73.704545   84.500000  443.409091  
Fairy      65.705882  78.529412  84.705882   48.588235  413.176471

In [152]:
df['Count'] = 1

df.groupby(['Type 1']).count()['Count']

Type 1
Bug          69
Dark         31
Dragon       32
Electric     44
Fairy        17
Fighting     27
Fire         52
Flying        4
Ghost        32
Grass        70
Ground       32
Ice          24
Normal       98
Poison       28
Psychic      57
Rock         44
Steel        27
Water       112
Name: Count, dtype: int64

In [156]:
df.groupby(['Type 1', 'Type 2']).count()['Count']

Type 1  Type 2  
Bug     Electric     2
        Fighting     2
        Fire         2
        Flying      14
        Ghost        1
                    ..
Water   Ice          3
        Poison       3
        Psychic      5
        Rock         4
        Steel        1
Name: Count, Length: 136, dtype: int64

# Working with Large Amounts of Data

In [160]:
big_df = pd.DataFrame(columns = df.columns)

for df in pd.read_csv('modified_pokemon.csv', chunksize = 5):
    chunks = df.groupby(['Type 1']).count()
    
    big_df = pd.concat([big_df, chunks])
    
print(big_df)

         # Name Type 1 Type 2 Generation Legendary HP Attack Defense Sp. Atk  \
Fire     1    1    NaN      0          1         1  1      1       1       1   
Grass    4    4    NaN      4          4         4  4      4       4       4   
Fire     4    4    NaN      3          4         4  4      4       4       4   
Water    1    1    NaN      0          1         1  1      1       1       1   
Bug      2    2    NaN      0          2         2  2      2       2       2   
...     ..  ...    ...    ...        ...       ... ..    ...     ...     ...   
Fairy    1    1    NaN      0          1         1  1      1       1       1   
Flying   2    2    NaN      2          2         2  2      2       2       2   
Fire     1    1    NaN      1          1         1  1      1       1       1   
Psychic  2    2    NaN      2          2         2  2      2       2       2   
Rock     2    2    NaN      2          2         2  2      2       2       2   

        Sp. Def Speed Total  
Fire     